js/modules/k6/execution/execution.go

```diff
@@ -0,0 +1,94 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package execution
+
+import (
+	"context"
+	"errors"
+
+	"github.com/loadimpact/k6/js/internal/modules"
+	"github.com/loadimpact/k6/lib"
+)
+
+func init() {
+	modules.Register("k6/execution", New())
+}
+
+// Execution is a JS module to return information about the execution in progress.
+type Execution struct{}
+
+// New returns a pointer to a new Execution.
+func New() *Execution {
+	return &Execution{}
+}
+
+// GetVUStats returns information about the currently executing VU.
+func (e *Execution) GetVUStats(ctx context.Context) (map[string]interface{}, error) {
+	vuState := lib.GetState(ctx)
+	if vuState == nil {
+		return nil, errors.New("VU information can only be returned from an exported function")
+	}
+
+	out := map[string]interface{}{
+		"id":        vuState.Vu,
+		"iteration": vuState.Iteration,
+	}
+
+	return out, nil
+}
+
+// GetScenarioStats returns information about the currently executing scenario.
+func (e *Execution) GetScenarioStats(ctx context.Context) (map[string]interface{}, error) {
+	ss := lib.GetScenarioState(ctx)
+	if ss == nil {
+		return nil, errors.New("scenario information can only be returned from an exported function")
+	}
+
+	progress, _ := ss.ProgressFn()
+	out := map[string]interface{}{
+		"name":      ss.Name,
+		"executor":  ss.Executor,
+		"startTime": ss.StartTime,
+		"progress":  progress,
+	}
```

Comment:
I think it will be better if `progress` calls the method when accessed (or is a function) because:
1. this won't call `progressFn` unless the user actually needs it instead of needing the `name` for example
2. will allow users to monitor the progress without calling `getScenarioStats` constantly

Reply:
Yeah, it would be a slight performance benefit to evaluate it on access. I'll look into doing this with `DynamicObject` then.

Reply:
I would say that a normal Object with `defineProperty` will be fine, but a `DynamicObject` will do the job as well ... just a little too much IMO

Reply:
Resolved by c034057e?